libfwupdplugin/fu-fmap-firmware.c

```diff
@@ -0,0 +1,258 @@
+{
+	gsize i;
+	gboolean fmap_found = FALSE;
+
+	if (offset == NULL) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "offset return not valid");
+		return FALSE;
+	}
+
+	for (i = 0; i < len - strlen(FMAP_SIGNATURE); i++) {
+		if (!memcmp(&image[i],
+		            FMAP_SIGNATURE,
+		            strlen(FMAP_SIGNATURE))) {
+			fmap_found = TRUE;
+			break;
+		}
+	}
+
+	if (!fmap_found) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "fmap not found using linear search");
+		return FALSE;
+	}
+
+	if (i + fmap_size ((FuFmap *)&image[i]) > len) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "malformed fmap too close to end of image");
+		return FALSE;
+	}
+
+	*offset = i;
+	return TRUE;
+}
+
+/* if image length is a power of 2, use binary search */
+static gboolean
+fmap_bsearch (const guint8 *image, gsize len, gsize *offset, GError **error)
+{
+	gsize i;
+	gboolean fmap_found = FALSE;
+
+	if (offset == NULL) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "offset return not valid");
+		return FALSE;
+	}
+
+	/*
+	 * For efficient operation, we start with the largest stride possible
+	 * and then decrease the stride on each iteration. Also, check for a
+	 * remainder when modding the offset with the previous stride. This
+	 * makes it so that each offset is only checked once.
+	 */
+	for (gint stride = len / 2; stride >= 1; stride /= 2) {
+		if (fmap_found)
+			break;
+
+		for (i = 0; i < len - strlen(FMAP_SIGNATURE); i += stride) {
+			if ((i % (stride * 2) == 0) && (i != 0))
+				continue;
+			if (!memcmp (&image[i],
+				     FMAP_SIGNATURE,
+				     strlen (FMAP_SIGNATURE))) {
+				fmap_found = TRUE;
+				break;
+			}
+		}
+	}
+
+	if (!fmap_found) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "fmap not found using binary search");
+		return FALSE;
+	}
+
+	if (i + fmap_size ((FuFmap *)&image[i]) > len)
+		return FALSE;
+
+	*offset = i;
+	return TRUE;
+}
+
+static gint
+popcnt (guint u)
+{
+	gint count;
+
+	/* K&R method */
+	for (count = 0; u; count++)
+		u &= (u - 1);
+
+	return count;
+}
+
+static gboolean
+fmap_find (const guint8 *image, gsize image_len, gsize *offset, GError **error)
+{
+	if (image == NULL || image_len == 0) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "invalid image");
+		return FALSE;
+	}
+
+	if (popcnt (image_len) == 1) {
+		if (!fmap_bsearch (image, image_len, offset, error)) {
+			g_prefix_error (error, "failed fmap_find using bsearch: ");
+			return FALSE;
+		}
+	} else {
+		if (!fmap_lsearch (image, image_len, offset, error)) {
+			g_prefix_error (error, "failed fmap_find using lsearch: ");
+			return FALSE;
+		}
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_fmap_firmware_parse (FuFirmware *firmware,
```

Comment:
Since this firmware can subclass, you need to call into potentially subclassed parse methods for it.  Example of how another plugin does this: https://github.com/fwupd/fwupd/blob/master/plugins/thunderbolt/fu-thunderbolt-firmware.c#L393

And yes that means in the subclassed one you don't call the superclassed parse (it's the other way around).
And from the cros-ec plugin you would just call `fu_firmware_parse` .  Another example here: https://github.com/fwupd/fwupd/blob/master/plugins/thunderbolt/fu-thunderbolt-device.c#L484

Reply:
I think the way it's done here is fine for the moment IMHO. If/when we get a subclassed version we can rework how the parse is handled.

Reply:
Thanks for the guidance. I've gone ahead and reworked this so that the subclass's parse just does the extra step of figuring out which EC specific image contains the section version strings and attaches them to the corresponding images. Much cleaner. Thanks!